### PR TITLE
Social Links: Add ability to change social icon sizes

### LINKED ISSUE #17276: Social Links Block: Add sizing options
In #16897 a "social links" block is likely to land in its first version. 

Based on feedback for that first version, we'd like to explore some additional features. 

One such feature is to add a size control to the sidebar, to give more control over how big the icons are. Such a size control should probably share DNA with the font size picker and suggest preset sizes rather than have a pixel slider, as the icons inside render best at specific sizes, usually even pixel numbers and multiples of .5x.

### PR DESCRIPTION
**Pull request by mkaz**
## Description

This adds a size button control to the block toolbar, that lets you pick between 4 different sizes of icons (Small, Normal, Large, Huge). Those sizes affect the default style, the pill shape style, the logos only style, and even those variations of the new setup state as well.

Fixes #17276

## How has this been tested?

- (Backwards Compatibility) Create a post with social icons and confirm after applying the PR that they still can be edited and changed.
- Add social icons, adjust their sizes using the sidebar
- Confirm the new sizes look right in the editor and published.

## Screenshots

social-link-size

## Types of changes

- Adds new attribute for size for Social Links
- Adds CustomSelect control to adjust sizes
- Adds custom CSS property for different sizes

### CODE REVIEW

**Comment by jasmussen** [comment]
This is what I see:

socl

Nice work! This will be a nice feature.

As you can see from my inspecting, it needs a little visual love that I can help with — more adaptive centering, to be specific, so as to work with arbitrary sizes.

One larger question, though, and I'll defer to folks like @nosolosw @kjellr @iamtakashi @alaczek and others: should this be a global CSS variable for all social links icons? Or should it be a per-block size setting? I've no real strong opinion but would love insights.

**Comment by nosolosw** [comment]
:wave: Took a quick look at this and it seems that you may be able to go with preset sizes (like small/medium/large) for the icons at the moment, is that correct? If so, I'd suggest making this work with classes instead, as the font size control does for presets.

If you also wanted custom values provided by the user, that's a bit more entangled and there's no clear solution yet for this kind of setup (where we want to set a property with a custom value in the container that affects all children). The issue is kind of similar to this other PR https://github.com/WordPress/gutenberg/pull/25372 The closest we've gone to address that use case is with the link control. The way it works there is by adding to the container a CSS variable with the custom value as an inline style. However, this approach is problematic with `kses` as it filters out any CSS variable within the style property of an element so it won't work for users without `unfiltered_html` permissions (author role, for example). See https://github.com/WordPress/gutenberg/pull/25411/ and https://github.com/Automattic/wp-calypso/issues/45456. The path forward seems to be either teach `kses` to allow any CSS variable within inline styles or look for an alternative approach.

**Comment by kjellr** [comment]
> One larger question, though, and I'll defer to folks like @nosolosw @kjellr @iamtakashi @alaczek and others: should this be a global CSS variable for all social links icons?

I think a global size for all social links icons is adequate. As @nosolosw mentioned, I think a small/medium/large dropdown sort of setup (along with perhaps a custom value field) would be great.

**Comment by jasmussen** [comment]
A set of presets is a good idea as it allows us to pick presets that render the SVG icons well. Here are some thoughts:

Huge (48px/2x):

huge


_Probably too huge, probably won't be used that often. But doesn't hurt to have it. Maybe you want to make a statement._


Large (36px/1.5x):

large


Big (32px/1.33_x) — best if we can do without this one as it's not a good scaling integer:

32 maybe


Default (24px/1x):

default

Small (18px/.75x):

small 18


Tiny (12px/.5x):

tiny 12

__Perhaps too small, but it's a decent scaling integer_

What do you think?

**Comment by kjellr** [comment]
Those sizes look good! I agree we can probably lose the "Big" one. The "Tiny" one is likely too small to recommend to users, so I think it's fine to remove it from the presets.

**Comment by jasmussen** [comment]
This is what I see:

size

Nice work, this will make a bunch of people happy.

However a couple of things — this control doesn't really fit in the sidebar, in part because it's a primary action for the block. It should just be a popover that opens from the block toolbar, it can just be a text button like this:

Screenshot 2020-10-15 at 13 49 06

And the font size indication isn't really helping here, it seems fine to just do a menu like this:

Screenshot 2020-10-15 at 13 51 34

File: packages/block-library/src/social-link/editor.scss
Comment (jasmussen) [inline packages/block-library/src/social-link/editor.scss:4]: Is it appropriate to surfaces this as a CSS variable even if it's not part of global styles? @nosolosw 

I think I'd prefer we avoid CSS variables entirely unless it's directly tied to global styles. An alternative is to have CSS classes, `is-small`, `is-large`, `is-huge` instead, and then apply the sizing based on that.

**Comment by jasmussen** [comment]
Thanks so much for this PR, I know my feedback means a little extra work. But I think it will be worth it, and that doing it the right way will leave it without additional work necessary.

File: packages/block-library/src/social-link/editor.scss
Comment (kjellr) [reply packages/block-library/src/social-link/editor.scss:4]: My expectation was that at some point themes might be able to modify this via global sizes — similar to how they can provide font size presets. 

That may be overkill from a design tools perspective, but I figured it can't hurt to allow that.

Comment (jasmussen) [reply packages/block-library/src/social-link/editor.scss:4]: Theme-abillity through global styles is definitely a nice feature, and if that requires these to be CSS variables (and if this is the correct way to do that), then I'm all thumbs up from here.

Comment (nosolosw) [reply packages/block-library/src/social-link/editor.scss:4]: I'd probably go with raw values instead of CSS variables in this PR. There's nothing wrong with using CSS vars as they are here (it's actually a nice use case). My only concern is that people start using them to tweak social icons so they become public API and we can't modify how this works down the road when/if this is integrated in Global Styles proper.

For some context: so far, the styles that can be declared via Global Styles are a straightforward 1:1 mapping to some CSS property (eg: fontSize, backgroundColor, etc). The needs here are a bit different in that four things need to be coordinated at once (width/height of the link, width/height of the logo), which makes that behaviour more particular to this specific block and less as a general utility (unlike fontSize or backgroundColor). This is not to say it can't be done, but that I see a few challenges that merit its own separate discussion.

Comment (mkaz) [reply packages/block-library/src/social-link/editor.scss:4]: Yes, the idea using CSS variables was that a theme would be able to provide a custom value or preset.

**Comment by mkaz** [comment]
I've moved the controls to the toolbar, test it out and let me know what you think. 👍

**Comment by jasmussen** [comment]
Thank you, this is what I see:

size

This is definitely the right interaction, but the button doesn't look quite as fitting as I had hoped. I think we should still try it for now, but I may go back and change it to an icon separately, if I can find a good size icon.

There needs to be a checkmark on the right side of the selected size (normal being selected by default). There should be code for that in the more options dropdown:

Screenshot 2020-10-16 at 09 40 15

**Comment by jasmussen** [comment]
This is unlikely to make it to 5.6, correct? Asking because looking at the code, I'd like to refactor it a bit, both to address Andrés feedback about not using CSS variables, but also to make the padding scale more naturally to the size. Ideally there's only one uniform padding that automatically pads more for larger icon sizes and less for smaller ones (like a percentage or something that scales with a value). But in order for that to work smoothly with both circles and pill shapes, I'll need to be able to focus on it for a bit, and I won't have time to do that until Monday.

**Comment by mkaz** [comment]
@jasmussen I updated with the check icon and setting the `isAlternate` prop that gives the proper type of popup. Both of those changes should address what you had above.

I also rebased against the placeholder social links, but in the rebase I broke things and now have chubby icons. I figured you wanted to rework some of the CSS, and you'll end up doing it much better than me, I'll let you work your magic to get them right. Needed for both in editor and published view. Thanks!

**Comment by paaljoachim** [comment]
My test shows this:

Social-icons-size

Btw
Should there also be a setting in the sidebar with a size drop down? Having it in the toolbar and the sidebar.

**Comment by jasmussen** [comment]
I'm going to take a stab at bringing this home today!

Thanks for taking a look, Paal.

> Should there also be a setting in the sidebar with a size drop down? Having it in the toolbar and the sidebar.

I'd suggest no: redundancy is okay in places, but it's hard to get right. It always begs the question: if there are two ways to accomplish the same, which one is the right way? And in absence of the redundancy, the principle to evaluate against is whether the action is primary (should be in toolbar) or secondary (should be in sidebar), and I'd suggest this is definitely primary.

**Comment by jasmussen** [comment]
As I worked on this, and looking for a clear and scalable (for themes also) heuristic for what the padding is, I ended up with a basic system. 6x6 units. An icon is 4x4 units with 1unit of padding all around. So if an icon is 24px, then it has a 6px padding (24/6). 

That works nicely with ems —`font-size: 24px; width: 1em; height: 1em; padding: 0.25em;` — however it did surface one gotcha.

18/4 = 4.5. And half-pixels don't render well, which means if we went with 18px for the "small" icon size, icons wouldn't be centered. They'd have 5px top and left padding, and 4px bottom and right padding. 

So I changed the small size from 18 to 16. It looks alright:

Screenshot 2020-10-19 at 10 21 37

**Comment by jasmussen** [comment]
@mkaz we need for the checkmark to be present on the "normal" size, even if you haven't explicitly set the property. I.e. insert a social links block, do nothing, check the size, and you see this:

need a default

That is — if you have no property explicitly set, the checkmark should also show on "normal".

**Comment by jasmussen** [comment]
Just testing this EM approach with a few themes, and so far it seems alright:

2020 back and front:

Screenshot 2020-10-19 at 10 30 44

Screenshot 2020-10-19 at 10 30 51

2019:

Screenshot 2020-10-19 at 10 31 06

Screenshot 2020-10-19 at 10 31 16

That margin in 2019 in the editor issue, I believe, is present also in the main branch.

**Comment by jasmussen** [comment]
Got variations working. Frontend:

front

Editor:

editor

Before:

Screenshot 2020-10-19 at 10 54 46

**Comment by jasmussen** [comment]
Got the setup new state working, both with variations and size:

Screenshot 2020-10-19 at 11 47 56

I think this one is actually ready to review.

**Comment by jasmussen** [comment]
Forgot to mention, we need https://github.com/WordPress/gutenberg/pull/25921#issuecomment-711840866 fixed as well before this one can ship. Let me take a stab now.

File: packages/block-library/src/social-link/editor.scss
Comment (draganescu) [inline packages/block-library/src/social-link/editor.scss:19]: Now, this is a very specific number 😄

Comment (jasmussen) [reply packages/block-library/src/social-link/editor.scss:19]: Heh, yes. The math checks out for the ems.

When the font size is 24px, 0.6666666666666 (infinite 6es but there wasn't space to write them all out) comes to 16px, which is the padding the pill shape should have left and right.

Comment (kirilzh) [reply packages/block-library/src/social-link/editor.scss:19]: I think in this case you can do `calc((2/3) * 1em)`

File: packages/block-library/src/social-links/style.scss
Comment (kirilzh) [inline packages/block-library/src/social-links/style.scss:55]: for the font-sizes you could do with size variables, since they are all multiples of 8

**Comment by jasmussen** [comment]
Oh nice, thank you @mkaz! Let me address the other feedback!

File: packages/block-library/src/social-links/style.scss
Comment (jasmussen) [reply packages/block-library/src/social-links/style.scss:55]: I tend to avoid variables unless they are to be reused in a number of places, and especially in the style.scss files, as they are just meant for the frontend.

File: packages/block-library/src/social-link/editor.scss
Comment (jasmussen) [reply packages/block-library/src/social-link/editor.scss:19]: That's not bad! Let me try that, thank you.

**Comment by jasmussen** [comment]
Nice. The checkmark correctly shows up for normal as default, even if it isn't explicitly set:

Screenshot 2020-10-19 at 15 31 17

And with the nice `calc` trick for the padding, the padding computes to the correct padding, as opposed to relying on rounding:

Screenshot 2020-10-19 at 15 32 49

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/pull/25921/commits/d4b71b790fdb3df25dedf73927eb7dacd157b44e I pushed a fix that adds specificity to the margin classes for social icons. While this is not the prettiest fix, it is still overridable if a theme has a good reason to do so, and it fixes an issue that is present in 2019. Before:

Screenshot 2020-10-19 at 19 00 50

After:

Screenshot 2020-10-19 at 19 06 03

File: packages/block-library/src/social-links/block.json
Comment (mtias) [inline packages/block-library/src/social-links/block.json:10]: Does this attribute need the icon prefix or could it be just "size"?

File: packages/block-library/src/social-links/edit.js
Comment (mtias) [inline packages/block-library/src/social-links/edit.js:27]: why is the `block-` removed?

File: packages/block-library/src/social-links/edit.js
Comment (mtias) [inline packages/block-library/src/social-links/edit.js:64]: Mmm, why is this needed here? We should not have these kind of definitions in the blocks themselves but rely instead on the toolbar components handling it

File: packages/block-library/src/social-links/edit.js
Comment (jasmussen) [reply packages/block-library/src/social-links/edit.js:27]: The frontend markup has the following CSS classes applied:

`<li class="wp-social-link wp-social-link-wordpress wp-block-social-links has-large-icon-size is-style-pill-shape">`

In effort to simply get the frontend and editor closer, it made sense to me to keep the classes more closely to the frontend.

Comment (mtias) [reply packages/block-library/src/social-links/edit.js:27]: Yes, I'm confused about when the block prefix is added and when excluded

Comment (jasmussen) [reply packages/block-library/src/social-links/edit.js:27]: I cleaned up a reference to `wp-block-social-link` so all CSS now refers to only `wp-social-link`. I'm not quite sure why we have both those classes, I'll defer to @mkaz.

**Comment by paaljoachim** [comment]
This is what I see:
Screen Shot 2020-10-21 at 12 25 18

Size selected is Large. Style: Default. 


Frontend:
Screen Shot 2020-10-21 at 12 26 12

Default and Pill Shape both show the above CSS styling.

File: packages/block-library/src/social-links/edit.js
Comment (nosolosw) [reply packages/block-library/src/social-links/edit.js:27]: FWIW, the `wp-block-social-links` is only in children (li elements) because of a bug we just fixed https://github.com/WordPress/gutenberg/pull/26344

**Comment by jasmussen** [comment]
_shudder_ — those circles are ovals! 🙈  — I'll take a look.

**Comment by jasmussen** [comment]
@paaljoachim I can't reproduce. This is what I see in twenty twenty:

Screenshot 2020-10-21 at 13 16 22

Can you try and make sure your cache is thoroughly cleared? Or are there other steps to reproduce?

**Comment by paaljoachim** [comment]
I updated npm in terminal by adding the command: `npm install`
Restarted terminal and Docker.
Then added the following:
`git fetch origin pull/25921/head`
`git checkout 5aef47571ce2bc6cfae8bb17d2fe721b758ecf15`
`npm run build`

Restarted and updated:
`wp-env start --update`

(Also run `git reset --hard`)

Opened http://localhost:8888/ in Chrome. Did a hard refresh of Chrome.

If the newest is supposed to show by default a checkmark next to the normal size, then I am for whatever reason still seeing an older commit. Which would probably explain the oval circles.

**Comment by jasmussen** [comment]
Two ways to make sure you're seeing the latest:
```
git checkout master
git pull
git branch -D [branch name]
git checkout [branch name]
npm install
npm run dev
```

then clear cache in your browser.

**Comment by paaljoachim** [comment]
The newest commit is working properly.

Backend:
Screen Shot 2020-10-21 at 14 37 50

Frontend:
Screen Shot 2020-10-21 at 14 38 34

**Comment by jasmussen** [comment]
Phew! Thanks!

File: packages/block-library/src/social-links/edit.js
Comment (mkaz) [reply packages/block-library/src/social-links/edit.js:64]: The event is fired when pressing the down arrow on the button, this is not the normal behavior for a button, but is natural (expected) in this case since it is opening a Dropdown menu, it will open and continuing down arrow will select the items.

I can see the need for a new type of button here that includes a dropdown, this behavior is also used in the Heading block for selecting level.

Testing again, and now it doesn't work as expected. The down arrow does open the menu, but pressing again does not move the selector, it loses focus.

File: packages/block-library/src/social-links/edit.js
Comment (mkaz) [reply packages/block-library/src/social-links/edit.js:27]: Marks this as resolved, per Andres comment.

File: packages/block-library/src/social-links/block.json
Comment (mkaz) [reply packages/block-library/src/social-links/block.json]: Switched by Joen, thanks Joen!

File: packages/block-library/src/social-links/edit.js
Comment (mtias) [reply packages/block-library/src/social-links/edit.js:64]: We have many controls in the toolbar that operate like dropdowns, seems it should be just handled with the core components. cc @diegohaz if you can give it a quick look

**Comment by jasmussen** [comment]
Should we try changing the "Size" label in the toolbar to show what's selected? Since it's essentially a dropdown just like Alignments, and alignments show the alignment chosen, we could do the same here. We'd just want to change "Normal" to "Medium".

**Comment by paaljoachim** [comment]
Yeah. I was thinking the same thing, but became hesitant that the selected size word would be seen and not the word "Size".

Here is an example of what it would look like.
Social-Icons-Block-Selected-size

File: packages/block-library/src/social-links/edit.js
Comment (diegohaz) [reply packages/block-library/src/social-links/edit.js:64]: @mkaz The generic `Dropdown` component doesn't add the key down handler automatically. `DropdownMenu` does. Have you tried it? For reference, the block switcher button uses it: https://github.com/WordPress/gutenberg/blob/99d02c51487c992699eab87c1c7ace90284d60e0/packages/block-editor/src/components/block-switcher/index.js#L159-L178

In case it helps, there's also this simple example on the developer handbook: https://developer.wordpress.org/block-editor/components/toolbar-item/#render-prop

Comment (mkaz) [reply packages/block-library/src/social-links/edit.js]: Thanks @diegohaz, I've got it working using `DropdownMenu` but some of the same issues persist with that component, also seen elsewhere, so I created a ticket to track. See #26401

**Comment by mkaz** [comment]
@jasmussen I updated using the DropdownMenu component, there are still a few issues related to that component created a tracking ticket here #26401 — the biggest change for this ticket is that the DropdownMenu component only allows a button and not text in the toolbar.

Joen, try it out and let me know what you think, I'm not sure if there is a good size icon to use, otherwise we can revert to the previous and address after the base components improve.

**Comment by jasmussen** [comment]
> the biggest change for this ticket is that the DropdownMenu component only allows a button and not text in the toolbar.

Screenshot 2020-10-23 at 09 27 29

Thanks so much for the work, but I can't see this working as well as the label. To me it's about this being a primary action for the block, therefore deserving not just a bigger button, but a label.

Is there any way we can extend or refactor the DropdownMenu to allow a label? Otherwise, what does the Image "Replace" button do to enable a dropdown from the label?

**Comment by mtias** [comment]
I think the image replace dropdown was a custom implementation and it has some keyboard issues still.

**Comment by jasmussen** [comment]
Argh. I just don't thin there's a good enough visual concept for "Size" to work as an icon.

**Comment by mkaz** [comment]
@joen @mtias I pushed up a change to test out a modification to `DropdownMenu` and `Button` to accept a `text` field and if `icon` is specified as null, the text will show instead of an icon.

This would be moved to its own PR but pushed up here as a way to test and try out.

**Comment by jasmussen** [comment]
Nice work. This is what I see:

social

Conceptually, that's perfect for me, except for two things:

- Small, normal, and large appear to not work anymore
- The dropdown menu itself should have the `isAlternate` prop to ensure the dark borders.

**Comment by jasmussen** [comment]
I rebased this to see if it will sate the e2e tests.

I'm seeing a thing in the branch at the moment, specifically that the sizes don't change when I choose a different one. It appears we have duplicate size classes now:

switch

Note how the initial saved class stays, and an additional class is added.

Finally, it really would be nice to show the sizes as small, medium, large, huge, directly in the button label. With those two fixed, it will be an awesome PR.

**Comment by jasmussen** [comment]
Nice, this is working well for me:

social links

One more change and I think this is good to go.

Labels should be Small, Medium, Large, Huge, and I think we should make them show up as the label in the toolbar. Is this workable?

**Comment by mapk** [comment]
Normally we've had the popover selection reflected in the block's toolbar (ie. alignment, position of text in Cover block, inline text color, etc ). I'm guessing representing the size with an icon is too ambiguous and showing the text of the size gets too wordy here?

**Comment by mkaz** [comment]
@jasmussen Updating with label in button. See gif or try on the PR.

@mapk does this address your comment too?

social-link-size-button

**Comment by mapk** [comment]
>@mapk does this address your comment too?

@mkaz thanks for trying it out so quickly! 
Showing the size in the toolbar, without any relationship to the other sizes, doesn't seem all that helpful IMO now that I'm seeing it in the gif.
I've also noticed that our text labels in the toolbar often indicate an action. "Large" or the other labels are descriptions. "Size" works pretty well because I really can't think of a better alternative. :) 
Sorry I'm not more helpful here. As you had it is probably best. I'll think on this more though. 👍

**Comment by jasmussen** [comment]
Ugh, it seems that the popover component centers itself according to the width of the toolbar control (which makes sense if you think about it), but that means that any time you change a size property and the toolbar updates, the popover shifts position.

In absence of being able to fix that so the popover stays still, let's go back to the toolbar saying "Size". Then we can revisit if/when it becomes pressing in the future. 

Let's keep the labels as small, medium, large, huge. "Normal" is in the eye of the beholder anyway!

**Comment by mkaz** [comment]
> Let's keep the labels as small, medium, large, huge. "Normal" is in the eye of the beholder anyway!

I initially was going to go with Medium, but looking at the font control it uses the "Normal" label, so for consistency I matched it, fine with changing either way.

**Review by jasmussen — APPROVED** [review]
One issue remains. Transforms are on two rows in this PR:

Screenshot 2020-11-06 at 16 05 23

However a rebase fixes that, and the issue is not present in the master branch:

Screenshot 2020-11-06 at 16 09 06

I suspect that an issue was created and fixed in the master branch, while this branch was ongoing.

But since rebasing will likely cause those beautiful green checks to fail a random amount of times, I'd say it's okay to merge this one if you first test whether you see the same issue as me, and that a rebase (even if not pushed) fixes it.

**Pull request merged**
